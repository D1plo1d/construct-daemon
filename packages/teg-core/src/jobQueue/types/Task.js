import { Record, List } from 'immutable'

import { SPOOLED_TASK } from './TaskStatusEnum'

// export type TaskT = RecordOf<{
//   id: string,
//   jobID: ?string,
//   jobFileID: ?string,

//   priority: PriorityEnumT,
//   internal: boolean,
//   isPollingRequest: boolean,
//   data: ListT<string>,
//   name: ?string,

//   /*
//    * The line number that will be executed by the printer by a saga after the
//    * reducers run.
//    */
//   currentLineNumber: ?number,

//   createdAt: ?number,
//   startedAt: ?number,
//   stoppedAt: ?number,
//   status: TaskStatusEnumT,
// }>

const TaskRecord = Record({
  id: null,
  machineID: null,
  jobID: null,
  jobFileID: null,
  name: null,
  filePath: null,
  commands: null,
  annotations: List(),
  totalLines: null,
  machineOverride: false,
  previousLineNumber: null,
  currentLineNumber: null,
  status: SPOOLED_TASK,
  onComplete: () => {},
  onError: () => {},
})

let nextTaskID = 1

const Task = (attrs) => {
  const {
    name,
    commands,
  } = attrs

  if (typeof name !== 'string') {
    throw new Error('name must be a string')
  }
  if (commands != null) {
    if (!Array.isArray(commands)) {
      throw new Error('commands must be an array of strings')
    }
  } else if (typeof attrs.filePath !== 'string') {
    throw new Error('One of task.commands or task.filePath must be set')
  }

  const id = attrs.id || nextTaskID
  nextTaskID += 1

  return TaskRecord({
    id,
    createdAt: new Date().toISOString(),
    ...attrs,
    commands,
  })
}

export const MockTask = attrs => Task({
  name: 'test.ngc',
  internal: true,
  data: ['g1 x10', 'g1 y20'],
  ...attrs,
})

export default Task
