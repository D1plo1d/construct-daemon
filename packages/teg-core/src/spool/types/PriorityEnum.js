import { List } from 'immutable'

export const EMERGENCY = 'EMERGENCY'
export const PREEMPTIVE = 'PREEMPTIVE'
export const NORMAL = 'NORMAL'

export type PriorityEnumT =
  | 'EMERGENCY'
  | 'PREEMPTIVE'
  | 'NORMAL'

export const priorityOrder = List([
  EMERGENCY,
  PREEMPTIVE,
  NORMAL,
])
