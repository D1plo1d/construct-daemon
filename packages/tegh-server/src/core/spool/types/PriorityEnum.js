export type PriorityEnumT =
  | 'emergency'
  | 'preemptive'
  | 'normal'

export const priorityOrder = List([
  'emergency',
  'preemptive',
  'normal',
])
