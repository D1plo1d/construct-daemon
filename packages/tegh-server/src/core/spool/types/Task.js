import type { RecordOf, List, Set } from 'immutable'
import type { Priority } from './priority'
import normalizeGCodeLines from '../helpers/normalize_gcode_lines'

export type TaskT = RecordOf<{
  id: string,
  priority: Priority,
  internal: boolean,
  data: List<string>,
  jobID: ?string,
  fileName: ?string,
  currentLineNumber: ?number,
  createdAt: ?number,
  startedAt: ?number,
  stoppedAt: ?number,
  status: 'initialized' | 'queued' | 'printing' | 'errored' | 'cancelled' | 'done',
}>

import uuid from 'uuid/v4'
import { Record, List } from 'immutable'

import normalizeGCodeLines from '../helpers/normalize_gcode_lines'

const TaskRecord = Record({
  id: null,
  priority: null,
  internal: null,
  fileName: null,
  jobID: null,
  data: null,
  status: null,
  currentLineNumber: null,
  createdAt: null,
  startedAt: null,
  stoppedAt: null,
})

const Task = ({
  priority,
  internal,
  fileName,
  data,
}) => TaskRecord({
  id: uuid(),
  createdAt: new Date().toISOString(),
  data: List(normalizeGCodeLines(data)),
  status: 'queued',
  priority,
  internal,
  fileName,
})

export default Task
