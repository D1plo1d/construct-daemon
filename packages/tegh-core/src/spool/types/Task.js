import type { RecordOf, List as ListT } from 'immutable'
import uuid from 'uuid'
import { Record, List } from 'immutable'
import type { PriorityEnumT } from './PriorityEnum'
import type { TaskStatusEnumT } from './TaskStatusEnum'


import normalizeGCodeLines from '../../util/normalizeGCodeLines'
import { priorityOrder, NORMAL } from './PriorityEnum'
import { SPOOLED } from './TaskStatusEnum'

export type TaskT = RecordOf<{
  id: string,
  jobID: ?string,
  jobFileID: ?string,

  priority: PriorityEnumT,
  internal: boolean,
  data: ListT<string>,
  name: ?string,

  /*
   * The line number that will be executed by the printer by a saga after the
   * reducers run.
   */
  currentLineNumber: ?number,

  createdAt: ?number,
  startedAt: ?number,
  stoppedAt: ?number,
  status: TaskStatusEnumT,
}>

const TaskRecord = Record({
  id: null,
  priority: null,
  internal: null,
  name: null,
  jobID: null,
  jobFileID: null,
  data: null,
  status: null,
  currentLineNumber: null,
  createdAt: null,
  startedAt: null,
  stoppedAt: null,
})

const Task = (attrs) => {
  const {
    priority,
    internal,
    name,
    data,
  } = attrs

  if (typeof name !== 'string') {
    throw new Error('name must be a string')
  }
  if (typeof internal !== 'boolean') {
    throw new Error('internal must be a boolean')
  }
  if (!priorityOrder.includes(priority)) {
    throw new Error('invalid priority')
  }
  if (!Array.isArray(data)) {
    throw new Error('data must be an array of strings')
  }
  return TaskRecord({
    id: uuid.v4(),
    createdAt: new Date().toISOString(),
    status: SPOOLED,
    ...attrs,
    data: List(normalizeGCodeLines(data)),
  })
}

export const MockTask = attrs => Task({
  name: 'test.ngc',
  internal: true,
  priority: NORMAL,
  data: ['g1 x10', 'g1 y20'],
  ...attrs,
})

export default Task
