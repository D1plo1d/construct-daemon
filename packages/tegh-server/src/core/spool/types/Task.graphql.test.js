import snapshotTestGraphQLType from '../../util/testing/snapshotTestGraphQLType'

import Task from './Task'
import { NORMAL } from './PriorityEnum'

import TaskGraphQL from './Task.graphql.js'

const task = Task({
  name: 'test task',
  internal: false,
  priority: NORMAL,
  data: ['g1 x10', 'g1 y20'],
})

const state = {
  config: {
    id: 'test_printer_id',
  },
}

snapshotTestGraphQLType('TaskGraphQL', {
  type: TaskGraphQL,
  fieldConfigs: {
    id: {
      dynamic: true,
    },
    createdAt: {
      dynamic: true,
    },
    percentComplete: {
      args: { digits: 5 },
    },
  },
  rootValue: task,
  contextValue: {
    store: {
      getState: () => state,
    },
  },
})
