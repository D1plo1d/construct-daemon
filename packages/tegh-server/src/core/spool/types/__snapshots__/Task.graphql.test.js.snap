// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TaskGraphQL presents a consistent API 1`] = `
Object {
  "description": "A spooled set of gcodes to be executed by the printer",
  "enumValues": null,
  "fields": Array [
    Object {
      "args": Array [],
      "deprecationReason": null,
      "description": null,
      "isDeprecated": false,
      "name": "id",
      "type": Object {
        "kind": "NON_NULL",
        "name": null,
        "ofType": Object {
          "kind": "SCALAR",
          "name": "ID",
          "ofType": null,
        },
      },
    },
    Object {
      "args": Array [],
      "deprecationReason": null,
      "description": null,
      "isDeprecated": false,
      "name": "name",
      "type": Object {
        "kind": "NON_NULL",
        "name": null,
        "ofType": Object {
          "kind": "SCALAR",
          "name": "String",
          "ofType": null,
        },
      },
    },
    Object {
      "args": Array [],
      "deprecationReason": null,
      "description": null,
      "isDeprecated": false,
      "name": "currentLineNumber",
      "type": Object {
        "kind": "SCALAR",
        "name": "Int",
        "ofType": null,
      },
    },
    Object {
      "args": Array [],
      "deprecationReason": null,
      "description": null,
      "isDeprecated": false,
      "name": "totalLineNumbers",
      "type": Object {
        "kind": "NON_NULL",
        "name": null,
        "ofType": Object {
          "kind": "SCALAR",
          "name": "Int",
          "ofType": null,
        },
      },
    },
    Object {
      "args": Array [],
      "deprecationReason": null,
      "description": null,
      "isDeprecated": false,
      "name": "percentComplete",
      "type": Object {
        "kind": "NON_NULL",
        "name": null,
        "ofType": Object {
          "kind": "SCALAR",
          "name": "Float",
          "ofType": null,
        },
      },
    },
    Object {
      "args": Array [],
      "deprecationReason": null,
      "description": null,
      "isDeprecated": false,
      "name": "createdAt",
      "type": Object {
        "kind": "NON_NULL",
        "name": null,
        "ofType": Object {
          "kind": "SCALAR",
          "name": "Date",
          "ofType": null,
        },
      },
    },
    Object {
      "args": Array [],
      "deprecationReason": null,
      "description": null,
      "isDeprecated": false,
      "name": "startedAt",
      "type": Object {
        "kind": "SCALAR",
        "name": "Date",
        "ofType": null,
      },
    },
    Object {
      "args": Array [],
      "deprecationReason": null,
      "description": null,
      "isDeprecated": false,
      "name": "stoppedAt",
      "type": Object {
        "kind": "SCALAR",
        "name": "Date",
        "ofType": null,
      },
    },
    Object {
      "args": Array [],
      "deprecationReason": null,
      "description": null,
      "isDeprecated": false,
      "name": "status",
      "type": Object {
        "kind": "NON_NULL",
        "name": null,
        "ofType": Object {
          "kind": "SCALAR",
          "name": "String",
          "ofType": null,
        },
      },
    },
    Object {
      "args": Array [],
      "deprecationReason": null,
      "description": null,
      "isDeprecated": false,
      "name": "printer",
      "type": Object {
        "kind": "NON_NULL",
        "name": null,
        "ofType": Object {
          "kind": "OBJECT",
          "name": "Printer",
          "ofType": null,
        },
      },
    },
  ],
  "inputFields": null,
  "interfaces": Array [],
  "kind": "OBJECT",
  "name": "Task",
  "possibleTypes": null,
}
`;

exports[`TaskGraphQL resolves every field without error 1`] = `
Object {
  "createdAt": "2018-03-23T00:04:03.342Z",
  "currentLineNumber": null,
  "id": "b00ccde5-7fa9-4437-8b0c-9097db5305bc",
  "name": "test task",
  "percentComplete": 0,
  "printer": Object {
    "id": "[DYNAMIC FIELD OMITTED]",
  },
  "startedAt": null,
  "status": "SPOOLED",
  "stoppedAt": null,
  "totalLineNumbers": 2,
}
`;
